### PR TITLE
feat(ci): document changes as they merge instead of per release

### PR DESCRIPTION
**Pull request by Haroenv**
## Summary

The docs automation ran once per release, as a single 50-turn Claude call. It hit the turn limit regularly and threw away the work when it did — [run 32247853114](https://github.com/algolia/instantsearch/actions/runs/32247853114) ended with:

```
"terminal_reason":"max_turns", "num_turns":51, "subtype":"error_max_turns"
"duration_ms":339141   ← 5m39s of a 15-minute timeout
"total_cost_usd":3.24
```

Turns were the binding constraint, not time or money — it used 37% of its timeout budget. But the more expensive problem was that the step exited 1, so *Create branch* and *Create PR* were skipped: seven correct edits to `docs-new/doc/api-reference/widgets/chat/react.mdx`, pushed nowhere.

Looking at the target repository turned up more. `algolia/docs-new` ships `AGENTS.md`, `.claude/CLAUDE.md`, `.claude/skills/update-doc/`, `style-guide.md` and `.vale.ini` — a full authoring contract that this workflow never loaded, because Claude ran from the workspace root and a `CLAUDE.md` in a subdirectory isn't picked up.

## What changed

**Split the work by change instead of by release.** `docs-feature.yml` runs on each merged `feat`/`fix` while its diff is small and unambiguous, and accumulates onto a long-lived staging branch in docs-new. `docs-automation.yml` now sweeps that branch for gaps at release time and opens one draft PR. Both serialise on a single concurrency group.

docs-new is shared across Algolia products, so the branch is namespaced — **`instantsearch/staging`**, not something generic like `docs/next` that another team's automation would collide with. Both workflows read it from one `DOCS_BRANCH_NAME` env value, so the docs team can veto the name with a one-line change. Turn pressure came from batching a week of releases into one prompt — this removes the batching. It also matches docs-new's own `update-doc` skill, which is per-PR, and their *"Cover one concern per PR"* rule, which a release batch structurally can't satisfy.

Both workflows use a new `docs-agent` composite action:

| | |
| --- | --- |
| **Conventions load** | Claude runs with docs-new as the working directory (`--add-dir` for the InstantSearch source), so their `.claude/CLAUDE.md → @../AGENTS.md → @../style-guide.md` chain and skills load. Our prompts shrank to what's genuinely InstantSearch-specific — the cross-flavor changelog trap, flavor mapping, checkpointing. |
| **Verification + feedback loop** | Vale on changed files at `--minAlertLevel=warning`, JSX tag balance, new pages checked against `docs.json`/`config/*.json`, links filtered to touched files. Failures feed **one fix-up pass**, then re-verify. `generate:flavors` runs after. |
| **Allowlist actually matches** | The old rules used `Bash(git diff *)` — space-star, not the `Bash(git diff:*)` prefix form — so they matched nothing, including `check:links`, which has therefore probably never run. Three of four Bash calls in the last run were denied outright. |
| **Continuation instead of failure** | An exhausted turn budget becomes a checkpoint in `PROGRESS.md`. A partial result still ships as a draft PR marked incomplete, with instructions to finish it locally. |
| **Telemetry** | Turns, cost, duration, stop reason and denied tool calls per pass, in the step summary. Denied calls are what would have caught the broken allowlist on day one. |

## Guides, not just reference

Reference pages are the easy half: a new option means a new row in a known table. The half that rots is a guide that walks through a flow the change just altered and still describes the old one — plausible, confident, and invisible because nothing is *missing*. The InstantSearch guide surface is 52 topics, including a whole `going-further/chat-customization/` tree that the current weekly chat work lands squarely on.

`docs-guide-impact.sh` derives search terms from the changed source paths, greps the docs tree, and appends a ranked list of pages that mention them. Terms matching more than 40 pages are dropped as too generic to identify anything — without that filter `open` and `display` matched **690 of 1958 pages**; with it, the real release diff yields 12:

```
| Terms | Page |
| ---: | --- |
|     6 | doc/api-reference/widgets/chat/react.mdx
|     6 | doc/api-reference/widgets/chat/js.mdx
|     4 | doc/guides/.../chat-customization/welcome-screen/react.mdx
|     3 | doc/guides/.../chat-customization/welcome-screen/js.mdx
|     2 | doc/guides/.../chat-customization/ai-search-experience/{react,js}.mdx
…
_Ignored as too common to identify anything: chat (48 pages), index (550 pages),
 templates (79 pages), tools (62 pages), types (152 pages), version (196 pages)._
```

Both prompts now treat guide staleness as a first-class obligation, with a deliberately high bar for creating a guide (a new widget, or a capability with nowhere to live) and a **ban on restructuring existing guides unattended** — those get proposed under a `Flagged for humans` heading instead. The release sweep additionally does the cycle-level check no per-change run can: only it can see that three separate changes have collectively made a walkthrough wrong.

**Also found while mapping that surface:** the flavor convention in the prompt was wrong, inherited from the original. There are no `*.js.mdx` files in docs-new — the flavor is the *file name* (`doc/api-reference/widgets/chat/js.mdx`), and navigation entries are the path without the extension (`/doc/api-reference/widgets/chat/js`). Both prompts now describe the real layout and leave other platforms' pages (`android.mdx`, `flutter.mdx`, …) alone.

**Security.** The clone URL embeds `DOCS_REPO_PAT` in `docs-new/.git/config`; the agent can read it, and anything it writes gets committed and pushed. Its prompt now also contains contributor-authored PR titles from a public repo. The action strips the token from the remote for the duration of the run and restores it via an `EXIT` trap.

**Commit convention.** `AGENTS.md` says *"This repo is docs-only, so `docs:` carries no information. Never use it"* — and PRs squash-merge there, so the old hardcoded `docs: ${TITLE}` landed a malformed commit on their default branch every time. Titles are now validated against their rule (`feat`/`fix`/`chore`, optional lowercase scope, sentence case, no trailing period, ≤70 chars) with a fallback when the model's title doesn't conform.

## Rebased onto the App-token change (#7182)

That landed on the same file this PR rewrites, so it's adopted rather than merged around. Both workflows now mint a scoped installation token with `actions/create-github-app-token`, check docs-new out with `actions/checkout`, and author commits as the App's bot user — no `DOCS_REPO_PAT` anywhere.

It interacts with the credential-hiding fix below. `actions/checkout` persists its credential as an `http.<url>.extraheader` in `.git/config`, which the agent can read, so stripping the remote URL is no longer sufficient — the prepare script unsets that header after the branch lifecycle work, and the commit script restores credentials for the push alone. Installation tokens last an hour and both job timeouts stay inside that.

`docs-staging-checkout.sh` → `docs-staging-prepare.sh`, since it no longer clones anything.

## Release blocking (@shaejaz's point)

A docs PR only *opens* at release time, so at that instant everything on it is published. But the staging branch keeps accumulating while the PR waits for review, so it can grow commits documenting features from *after* that release — and merging then ships docs for an API nobody can install. Nothing tracked that.

Each staging commit now carries `Source-PR` / `Source-Commit` trailers. At release time the workflow extracts the PR numbers the release publishes from the changelogs, splits the branch into published and not, and maintains **one self-updating comment** on the docs PR (found by an HTML marker, so it's edited rather than re-posted every release):

```
⏳ Blocked. 1 change(s) documented here are not in a published release yet.
Merging now would document an API that cannot be installed.

### In a published release
- fix(ui-libraries): document the trigger message (algolia/instantsearch#7166)

### Not released yet
- feat(ui-libraries): document a brand new widget (algolia/instantsearch#7190)
```

Once the next release publishes the rest, the same comment flips to `✅ No longer blocked`. There's also an `unreleased` label so the state is filterable rather than only prose, and the PR is promoted out of draft once everything on it is released **and** the sweep finished **and** checks pass — matching docs-new's rule about draft-until-checks-green. Promote only, never demote: a human taking it out of draft deliberately shouldn't be undone.

## Review findings

All seven were real; two were things my own tests should have caught.

| Finding | Fix |
| --- | --- |
| **The push token was still reachable.** Replacing the remote URL hid it from `.git/config`, but the agent step's own environment still carried it — readable from `/proc/self/environ`. | The token never enters that step. The checkout script leaves the remote tokenless; the commit script supplies it for the push alone and restores the tokenless URL on a `trap`, so the failure path is covered too. |
| **Squash-merge detection was broken by construction.** docs-new squash-merges, which rewrites the commit, so the staging tip is *never* an ancestor of the default branch afterwards. The ancestor test concluded "still open" forever, dragging merged content into every later cycle. | Asks GitHub whether the branch's head was merged, rebuilds from the default branch, and cherry-picks anything pushed after the merge. |
| **The path boundary only warned**, then `git add -A` committed the violation anyway. | Refuses to push, and stages the content paths explicitly. |
| **A zero exit code was treated as completion.** `PROGRESS.md`'s marker is the contract; a model that stops early exits zero too. | A clean exit without the marker ends the loop without claiming the work is done. |
| **A failed `generate:flavors`** left the run reporting clean checks with possibly-stale navigation. | It's a check failure, and it lands in the report. |
| **Verification skipped everything already committed** — it ran only when the working tree was dirty, and used revision-less diffs. The release sweep could open a PR over content nothing had checked. | Covers `origin/<default>...HEAD` as well as the working tree. |
| **The fallback title was 74 characters**, breaking the 70-char rule it exists to enforce. | Shortened to 65, and the fallback is now length-checked itself. |

On the squash-merge one: my lifecycle test passed only because I simulated a fast-forward merge rather than the squash the repo actually uses. Reproducing it with `git commit-tree` shows the old check plainly returning "not merged".

## Testing

No CI covers these paths, so everything was exercised locally:

- **Composite action end-to-end** with stubbed `claude`/`npm`/`npx` against a throwaway docs repo. Injected an unbalanced `<Steps>` and an unregistered new page; both checks fired, the fix-up pass ran, re-verification came back clean, and the PAT was confirmed absent during the run and restored after.
- **Title validator** against five inputs — `docs:` prefix, Title Case, trailing period, and a 71-character title all fall back; the conventional one passes through.
- **Release context builder** run against this repository: 27 source files, 92 lines of context.
- **Guide impact** run against a real docs-new clone (1958 pages) with a real release diff and a single real commit — 12 pages each, with the relevant `chat-customization` guides ranked at the top; empty input exits cleanly.
- **Staging branch lifecycle** against a throwaway bare remote: fresh create, continue diverged, recreate after fast-forward merge, and — added for the review — recreate after a real squash merge, plus replay of a commit pushed after that merge.
- **Release status** against a fixture staging branch: blocked (label added, no promotion), unblocked (comment `PATCH`ed not re-posted, label removed, promoted), unblocked-but-checks-failing (no promotion), already-ready (no redundant call), and the sweep's own commit with no provenance trailer (counted as released). Changelog PR-number extraction checked against the real `instantsearch.js` changelog — 7 pull requests, correctly excluding the `closes [#6880]` issue ref.
- **Re-run after the rebase** against the new shape: a pre-checked-out docs-new with a planted `extraheader` credential, a stubbed `gh` for the default branch / bot user / merged-PR lookups. Confirms the credential count goes `1 → 0` across prepare, the commit is authored `fx-bot[bot] <12345+fx-bot[bot]@users.noreply.github.com>`, nothing is left behind after the push, and squash detection still fires.
- **Post-review paths**, each reproduced then re-run: verification catching a defect in already-committed staging content with zero working-tree edits; clean exit without a marker reporting `incomplete`; `generate:flavors` failure reporting `issues`; the boundary check refusing to push a file outside `doc/`; commit-and-push against a throwaway remote confirming the remote is left tokenless afterwards.
- One more bug found while testing the boundary fix: `git add -A -- <paths>` aborts entirely if any listed path is absent from the repo, so nothing was staged. It stages per path now.
- `bash -n` + `shellcheck -S warning` on every `run:` block and both scripts; `js-yaml` parse on all three YAML files.

Two bugs found and fixed while testing: `git diff` can't see untracked files, so **new pages escaped both the tag and navigation checks** — the navigation check only ever fires for new pages, so it was dead code (fixed with `git add -A -N`); and `git … | head` under `pipefail` SIGPIPEs into a failed step.

## Rollout notes

- **Staging branch lifecycle:** created from the default branch on first use, reused across a cycle, and recreated once its PR merges (detected via `git merge-base --is-ancestor`). All three paths tested functionally against a throwaway bare remote.
- **Cost:** roughly 10 small runs per week (many exiting immediately via the scope filter) plus a cheap release sweep, versus one $3–10 batch. Similar total, much better failure isolation.
- **`CLAUDE.md`** gains one line, in `Conventions`: a cross-repo agent workflow must run with cwd inside the target repo, and `--allowedTools` needs the `Bash(cmd:*)` prefix form. Those two mistakes account for four of the six defects here.
- **Worth testing manually first** via `workflow_dispatch` on `docs-automation.yml` with `create_pr: false` before letting the push trigger run.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .github/actions/docs-agent/action.yml [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+    value: ${{ steps.agent.outputs.verify }}
+
+runs:
+  using: composite
+  steps:
+    - id: agent
+      shell: bash
+      env:
+        ANTHROPIC_API_KEY: ${{ inputs.anthropic-api-key }}
+        DOCS_PAT: ${{ inputs.docs-pat }}
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply .github/actions/docs-agent/action.yml]: Fixed in e8c2a45. Correct, and the more important half of the problem — `/proc/self/environ` was still readable even with the remote URL scrubbed. The token no longer enters the agent step at all: `docs-staging-checkout.sh` leaves the remote tokenless, and `docs-staging-commit.sh` sets the push URL only for its own `git push`, restoring the tokenless URL on a `trap` so the failure path is covered too. Verified by running the commit script against a throwaway remote and confirming no credential in `.git/config` afterwards, including when the push fails.

_— via Claude_

File: .github/scripts/docs-staging-prepare.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+git config user.name "github-actions[bot]"
+git config user.email "github-actions[bot]@users.noreply.github.com"
+
+DEFAULT_BRANCH=$(git symbolic-ref --short refs/remotes/origin/HEAD | sed 's#^origin/##')
+echo "Default branch: ${DEFAULT_BRANCH}"
+
+RECREATED=false
+if git ls-remote --exit-code --heads origin "$BRANCH" >/dev/null 2>&1; then
+  git fetch --quiet origin "$BRANCH"
+  if git merge-base --is-ancestor "origin/${BRANCH}" "origin/${DEFAULT_BRANCH}"; then
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply .github/scripts/docs-staging-prepare.sh:55]: Fixed in e8c2a45. This one was broken by construction, and my own lifecycle test missed it because I simulated a fast-forward merge rather than the squash the repo actually uses. It now asks GitHub whether the branch head was merged (`gh pr list --head … --state merged --json headRefOid`), rebuilds from the default branch, and cherry-picks anything pushed after the merge. Reproduced with `git commit-tree` first, which shows the old check plainly returning "not an ancestor".

_— via Claude_

File: .github/actions/docs-agent/action.yml [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+              echo "Fix-up pass resolved the issues."
+            else
+              echo "::warning::Checks still failing after the fix-up pass."
+              cat "$STATE/verify-report.md"
+            fi
+          fi
+
+          # Flavor switchers are generated, not authored
+          ( cd "$DOCS" && npm run --silent generate:flavors ) || \
+            echo "::warning::generate:flavors failed"
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply .github/actions/docs-agent/action.yml]: Fixed in e8c2a45. A `generate:flavors` failure now appends to the verification report and sets `VERIFY=issues`, so the draft PR surfaces it instead of claiming clean checks over possibly-stale navigation.

_— via Claude_

File: .github/scripts/docs-staging-commit.sh [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  exit 0
+fi
+
+# The agent has no business touching build config, CI, or its own instructions.
+UNEXPECTED=$(git status --porcelain | awk '{ print $NF }' \
+  | grep -vE '^(doc/|docs\.json|config/|snippets/|include-snippets/|images/)' || true)
+if [ -n "$UNEXPECTED" ]; then
+  echo "::warning::The agent touched files outside the docs content paths:"
+  printf '%s\n' "$UNEXPECTED"
+fi
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply .github/scripts/docs-staging-commit.sh:43]: Fixed in e8c2a45. The check now `exit 1`s with the offending paths instead of logging them, and staging is explicit (`git add` per content path) rather than `git add -A`. Fixing it surfaced a second bug: a single `git add -A -- <many paths>` aborts entirely if any listed path is absent from the repo, staging nothing at all — so it adds one path at a time now.

_— via Claude_

File: .github/actions/docs-agent/action.yml [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+            break
+          fi
+
+          if [ "$RC" -eq 0 ]; then
+            STATUS="complete"
+            echo "Agent exited cleanly without a STATUS marker; treating as complete."
+            break
+          fi
+
+          echo "::warning::Pass ${PASS} did not finish (exit ${RC}). Resuming from the checkpoint."
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply .github/actions/docs-agent/action.yml:290]: Fixed in e8c2a45, though not exactly as suggested. You're right that the marker is the contract and a clean exit doesn't prove completion, so the run is no longer reported as complete. But re-running a model that already decided it was finished just burns a pass, so a clean exit without the marker now *ends* the loop while reporting `incomplete` — the release PR still gets its warning.

_— via Claude_

File: .github/actions/docs-agent/action.yml [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+        done
+
+        # --- verify, then one fix-up pass ------------------------------------
+        VERIFY="clean"
+        if [ -n "$(git -C "$DOCS" status --porcelain)" ]; then
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply .github/actions/docs-agent/action.yml]: Fixed in e8c2a45. Both halves were real: the guard now also fires when the branch has commits beyond the base, and the changed/added file lists union the working tree with `origin/${DOCS_DEFAULT_BRANCH}...HEAD`. Regression-tested with zero working-tree edits and a broken page already committed on the staging branch — that now reports `verify=issues` where it previously reported `clean`.

_— via Claude_

File: .github/workflows/docs-automation.yml [hunk 7]
```diff
@@ -54,145 +72,258 @@ jobs:
...
-          echo ""
-          echo "Would run Claude with prompt to:"
-          echo "1. Read changelogs from instantsearch packages"
-          echo "2. Explore docs-new structure"
-          echo "3. Update documentation for new features/changes"
-          echo "4. Run link check to catch broken links"
-          echo ""
-          echo "Trigger with run_claude=true to actually run Claude."
+          DOCS_REPO_PAT: ${{ secrets.DOCS_REPO_PAT }}
+          DOCS_FALLBACK_TITLE: 'fix(ui-libraries): update the InstantSearch documentation for this release'
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply .github/workflows/docs-automation.yml]: Fixed in e8c2a45 — 74 characters confirmed. Both this one and the direct PR-title fallback at what was line 222 are now `fix(ui-libraries): update the InstantSearch docs for this release` (65). The fallback is length-checked at runtime too, since a fallback that breaks the rule it exists to enforce is worse than no check.

_— via Claude_

**Review by shaejaz — COMMENTED** [review]
I might have missed this if it's already mentioned, but maybe once a release is done for the PR that had been merged, it can leave a comment on the docs PR that it should no longer be blocked.
Might be a bit complicated to do, but would be helpful in keeping track.

**Comment by Haroenv** [comment]
Good catch, and it was a real hole rather than something already covered.

The docs PR only *opens* at release time, so at that instant everything on it is published. But the staging branch keeps accumulating while the PR sits in review, so by the time someone looks at it, it can have grown commits documenting features from *after* that release — and merging then ships docs for an API nobody can install. Nothing tracked that.

Implemented in b422232c0:

- Each staging commit now carries `Source-PR` / `Source-Commit` trailers, so provenance survives into the docs repo.
- At release time the workflow extracts the PR numbers this release publishes from the changelogs, splits the branch into published and not, and maintains **one self-updating comment** (found by an HTML marker, so it's edited rather than re-posted each release):

  > :hourglass: **Blocked.** 1 change(s) documented here are not in a published release yet.
  > Merging now would document an API that cannot be installed. This comment updates itself on the next release.
  >
  > **In a published release**
  > - fix(ui-libraries): document the trigger message (algolia/instantsearch#7166)
  >
  > **Not released yet**
  > - feat(ui-libraries): document a brand new widget (algolia/instantsearch#7190)

  and once the next release publishes the rest, the same comment flips to ":white_check_mark: **No longer blocked.**"

- Plus an `unreleased` label, so it's filterable and not just prose, and the PR is promoted out of draft once everything on it is released **and** the sweep finished **and** Vale/link checks pass — which matches docs-new's own rule about keeping PRs in draft until checks are green. Promote only, never demote: if a human takes it out of draft deliberately, that shouldn't be undone.

Not as complicated as feared — the trailers made it mostly bookkeeping. Tested against a fixture branch covering blocked, unblocked, unblocked-but-checks-failing, already-ready, and the sweep's own untagged commit.

_— via Claude_

**Review by anjola-adeuyi — APPROVED** [review]

File: .github/scripts/docs-staging-commit.sh [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+  # Provenance, so the release run can work out which of these documented
+  # changes are in a published release and which would document an API nobody
+  # can install yet. docs-release-status.sh reads these trailers back.
+  if [ -n "${SOURCE_PR:-}" ]; then
+    printf 'Source-PR: %s\n' "$SOURCE_PR"
+  fi
+  if [ -n "${SOURCE_COMMIT:-}" ]; then
+    printf 'Source-Commit: %s\n' "$SOURCE_COMMIT"
+  fi
+  printf '\nCo-Authored-By: Claude <noreply@anthropic.com>\n'
```
Comment (anjola-adeuyi) [inline .github/scripts/docs-staging-commit.sh:81]: `Source-PR` sits above the `Co-Authored-By` paragraph, so `%(trailers:key=Source-PR)` reads empty and the status comment always says `No longer blocked`. your call whether it moves into the last paragraph.
otherwise looks good to me

Comment (Haroenv) [reply .github/scripts/docs-staging-commit.sh:81]: I think that's fine
